### PR TITLE
Add Spectrogram CPU Operator

### PR DESCRIPTION
**Pull request by jantonguirao**
#### Why we need this PR?
- We need a Spectrogram operator

#### What happened in this PR?
- Implemented Spectrogram CPU operator based on existing kernels
- Added tests for it, comparing against librosa

**JIRA TASK**: [DALI-1162]

### CODE REVIEW

File: dali/operators/signal/fft/spectrogram.cc [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  .NumOutput(1)
+  .AddOptionalArg("nfft",
+    R"code(Size of the FFT. The number of bins created in the output is `nfft // 2 + 1`
+(positive part of the spectrum only).)code",
+    -1)
+  .AddOptionalArg("window_length",
+    R"code(Window size (in number of samples))code",
+    512)
+  .AddOptionalArg("window_step",
+    R"code(Step betweeen the STFT windows)code",
```
Comment (JanuszL) [inline dali/operators/signal/fft/spectrogram.cc]:
```suggestion
    R"code(Step betweeen the STFT windows (in number of samples))code",
```

File: dali/operators/signal/fft/spectrogram.cc [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+    256)
+  .AddOptionalArg("window_fn",
```
Comment (JanuszL) [inline dali/operators/signal/fft/spectrogram.cc]: Why not AdditionalInput?

File: cmake/Dependencies.common.cmake [hunk 3]
```diff
@@ -173,4 +173,4 @@ if (BUILD_FFTS)
   include_directories(SYSTEM ${PROJECT_SOURCE_DIR}/third_party/ffts/include)
   list(APPEND DALI_LIBS ffts)
   list(APPEND DALI_EXCLUDES libffts.a)
-endif()
+endif()
```
Comment (mzient) [inline cmake/Dependencies.common.cmake]: Don't.

Comment (jantonguirao) [reply cmake/Dependencies.common.cmake]: Done

File: dali/operators/signal/fft/spectrogram.cc (hunk 2 above)
Comment (mzient) [reply dali/operators/signal/fft/spectrogram.cc]: My guess is that it's not likely to be per-sample.

Comment (jantonguirao) [reply dali/operators/signal/fft/spectrogram.cc]: It's a small list of float32 numbers (nfft size) describing the function. The user will just bake it in python (or use scipy or similar to generate the function) and pass it to the operator as a python list.

File: dali/operators/signal/fft/spectrogram.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/operators/signal/fft/spectrogram.cc]: Done

**Comment by jantonguirao** [comment]
!build

File: dali/operators/signal/fft/spectrogram.cc [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        break;
+      default:
+        DALI_FAIL(make_string("`power` can be only 1 (energy) or 2 (power), received ", power));
+    }
+
+    // layout (.., freq, time)
+    args.transform_axis = ndims - 2;
+  }
+
+  void CreateWindowHann(span<float> window_fn_data) {
```
Comment (mzient) [inline dali/operators/signal/fft/spectrogram.cc]: I'd move it to `kernels/signal/window/window_functions.h` or something similar.

File: dali/operators/signal/fft/spectrogram.cc (hunk 2 above)
Comment (JanuszL) [reply dali/operators/signal/fft/spectrogram.cc]: I was asking the thing that @mzient just answered. Fair.

File: dali/test/python/test_operator_spectrogram.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        out = self.data.gpu() if self.device == 'gpu' else self.data
+        out = self.fft(out)
+        return out
+
+    def iter_setup(self):
+        data = self.iterator.next()
+        self.feed_input(self.data, data)
+
+def spectrogram_func_torch(nfft, win_len, win_step, input):
+    def hann_win(n):
```
Comment (JanuszL) [inline dali/test/python/test_operator_spectrogram.py]: Can we have this `hann_win` extracted to one place. Now I see it twice.

File: dali/test/python/test_operator_spectrogram.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    def __init__(self, device, batch_size, iterator, nfft, window_length, window_step,
+                 num_threads=1, device_id=0):
+        super(SpectrogramPythonPipeline, self).__init__(
+              batch_size, num_threads, device_id,
+              seed=12345, exec_async=False, exec_pipelined=False)
+        self.device = "cpu"
+        self.iterator = iterator
+        self.inputs = ops.ExternalSource()
+
+        function = partial(spectrogram_func_librosa, nfft, window_length, window_step)
```
Comment (JanuszL) [inline dali/test/python/test_operator_spectrogram.py]: I would make this `spectrogram_func_librosa` an argument as you can use `spectrogram_func_torch` there as well.

File: dali/test/python/test_operator_spectrogram.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    for device in ['cpu']:
+        for batch_size in [3]:
+            for nfft, window_length, window_step, length in [(256, 256, 128, 4096),
+                                                             (16, 16, 8, 1000),
+                                                             (10, 10, 5, 1000),
+                                                             ]:
+                yield check_operator_spectrogram_vs_python_wave_1d, device, batch_size, length, \
+                    nfft, window_length, window_step
+
+def debug_spectrogram(device, batch_size, data1, data2, nfft, window_length, window_step):
```
Comment (JanuszL) [inline dali/test/python/test_operator_spectrogram.py]: I wonder if we need this debug function included inside this test.

File: dali/test/python/test_utils.py [hunk 8]
```diff
@@ -164,3 +168,22 @@ def __next__(self):
         return (batch)
 
     next = __next__
+
+class DataIterator(object):
+    def __init__(self, batch_size, y_data, dtype):
+        self.batch_size = batch_size
+        self.test_data = []
+        for _ in range(self.batch_size):
+            self.test_data.append(np.array(y_data, dtype=dtype))
```
Comment (JanuszL) [inline dali/test/python/test_utils.py]: Do we really wan tot have the same data across the samples?

File: docs/examples/audio/spectrogram.ipynb [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/usr/local/lib/python2.7/dist-packages/librosa/__init__.py:35: FutureWarning: You are using librosa with Python 2. Please note that librosa 0.7 will be the last version to support Python 2, after which it will require Python 3 or later.\n",
```
Comment (JanuszL) [inline docs/examples/audio/spectrogram.ipynb]: Remove this debug output.

File: docs/examples/audio/spectrogram.ipynb [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+    "\n",
+    "Typically, and in our DALI implementation, a spectrogram is calculated by computing the fast fourier transform (FFT) over a series of overlapping windows extracted from the original signal. The process of dividing the signal in short term sequences of fixed size and applying FFT on those independently is called Short-time Fourier transform (STFT). The spectrogram is then calculated as the power of the complex magnitude of the STFT.\n",
+    "\n",
+    "Extracting short term windows of the original image affects the calculated spectrum by producing aliasing artifacts. This is often call spectral leakage. To control/reduce the spectral leakage effect, we use different window functions when extracting short term windows. Some examples of window functions are: Hann, rectangular, Hanning.\n",
+    "\n",
+    "It is beyond the scope of this example to go deeper into the details of the signal processing concepts we mentioned above."
```
Comment (JanuszL) [inline docs/examples/audio/spectrogram.ipynb]: Maybe you can provide some reference where the user can learn more.

File: docs/examples/audio/spectrogram.ipynb [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    "Librosa provides an API to calculate the STFT, producing a complex output (i.e. complex numbers). It is then trivial to calculate the power spectrum from the complex STFT by the following"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_fft=2048 \n",
```
Comment (JanuszL) [inline docs/examples/audio/spectrogram.ipynb]: I would describe what that arguments mean.

File: docs/examples/audio/spectrogram.ipynb [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+    "import nvidia.dali as dali\n",
+    "\n",
+    "class SpectrogramPipeline(Pipeline):\n",
+    "    def __init__(self, device, batch_size, nfft, window_length, window_step, num_threads=1, device_id=0):\n",
+    "        super(SpectrogramPipeline, self).__init__(batch_size, num_threads, device_id)\n",
+    "        self.device = device\n",
+    "        \n",
+    "        self.batch_data = []\n",
+    "        for _ in range(batch_size):\n",
+    "            self.batch_data.append(np.array(y, dtype=np.float32))\n",
```
Comment (JanuszL) [inline docs/examples/audio/spectrogram.ipynb]: It is hard to tell that this `y` comes from `librosa.load(librosa.util.example_audio_file())`.
I would again load this just to make it more verbose.

File: dali/operators/signal/fft/spectrogram.cc (hunk 2 above)
Comment (mzient) [reply dali/operators/signal/fft/spectrogram.cc]: I think we should have some utilities for generating common window functions - even if just in Python.

File: dali/operators/signal/fft/spectrogram.cc [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+  using FftKernel = kernels::signal::fft::Fft1DCpu<OutputType, InputType, WindowsDims>;
+
+  explicit SpectrogramImplCpu(const OpSpec & spec);
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const workspace_t<CPUBackend> &ws) override;
+  void RunImpl(workspace_t<CPUBackend> &ws) override;
+
+ private:
+  int nfft_ = -1;
+  int64_t window_length_ = -1;
+  int64_t window_step_ = -1;
```
Comment (mzient) [inline dali/operators/signal/fft/spectrogram.cc]: FYI - I'm changing these to `int` in `ExtractWindowsArgs` - I don't think it's practical to assume that window size or step would be larger than 2G elements.

File: dali/operators/signal/fft/spectrogram.cc [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+
+  std::vector<OutputDesc> window_out_desc_;
+  std::vector<Tensor<CPUBackend>> window_out_;
+
+  kernels::KernelManager kmgr_fft_;
+  kernels::signal::fft::FftArgs fft_args_;
+};
+
+namespace {
+  void FillExtractWindowsArgs(kernels::signal::window::ExtractWindowsArgs& args,
```
Comment (mzient) [inline dali/operators/signal/fft/spectrogram.cc]: There's very little logic in this function and it's used only once - maybe you can just use braced initializer at call site instead?

File: dali/test/python/test_operator_spectrogram.py (hunk 5 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_spectrogram.py]: This torch implementation was meant to be replaces by librosa (there is no need to check both as torchaudio has already test verifying their implementation against librosa).

File: dali/test/python/test_operator_spectrogram.py (hunk 6 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_spectrogram.py]: I removed spectrogram_func_torch, but I'll make it an argument anyway

File: dali/test/python/test_operator_spectrogram.py (hunk 7 above)
Comment (jantonguirao) [reply dali/test/python/test_operator_spectrogram.py]: removed

File: dali/test/python/test_utils.py (hunk 8 above)
Comment (jantonguirao) [reply dali/test/python/test_utils.py]: I renamed this as "ConstantDataIterator"

File: docs/examples/audio/spectrogram.ipynb (hunk 9 above)
Comment (jantonguirao) [reply docs/examples/audio/spectrogram.ipynb]: done

File: docs/examples/audio/spectrogram.ipynb (hunk 12 above)
Comment (jantonguirao) [reply docs/examples/audio/spectrogram.ipynb]: will do

File: docs/examples/audio/spectrogram.ipynb (hunk 11 above)
Comment (jantonguirao) [reply docs/examples/audio/spectrogram.ipynb]: Will do

File: docs/examples/audio/spectrogram.ipynb (hunk 10 above)
Comment (jantonguirao) [reply docs/examples/audio/spectrogram.ipynb]: Done

File: dali/operators/signal/fft/spectrogram.cc (hunk 13 above)
Comment (jantonguirao) [reply dali/operators/signal/fft/spectrogram.cc]: Changed

File: dali/operators/signal/fft/spectrogram.cc (hunk 14 above)
Comment (jantonguirao) [reply dali/operators/signal/fft/spectrogram.cc]: Removed

File: dali/kernels/signal/window/window_functions.h [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+namespace kernels {
+namespace signal {
+
+template <typename T>
+void HannWindow(span<T> output) {
+  int N = output.size();
+  double a = (2 * M_PI / N);
+  for (int t = 0; t < N; t++) {
+    double phase = a * (t + 0.5);
+    output.data()[t] = static_cast<T>(0.5 * (1.0 - std::cos(phase)));
```
Comment (klecki) [inline dali/kernels/signal/window/window_functions.h]: Why not operator[] directly?
```suggestion
    output[t] = static_cast<T>(0.5 * (1.0 - std::cos(phase)));
```
Same below.

**Comment by mzient** [comment]
!build

File: dali/kernels/signal/window/window_functions.h (hunk 15 above)
Comment (jantonguirao) [reply dali/kernels/signal/window/window_functions.h]: done

**Pull request merged**